rclcpp/include/rclcpp/rate.hpp

```diff
@@ -112,8 +114,86 @@ class GenericRate : public RateBase
   std::chrono::time_point<Clock, ClockDurationNano> last_interval_;
 };
 
-using Rate = GenericRate<std::chrono::system_clock>;
-using WallRate = GenericRate<std::chrono::steady_clock>;
+using WallRate = GenericRate<std::chrono::system_clock>;
+using SteadyRate = GenericRate<std::chrono::steady_clock>;
+using Rate [[deprecated("Use rclcpp::WallRate explicitly instead.")]] = WallRate;
+
+class ROSRate : public RateBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(ROSRate)
+
+  explicit ROSRate(rclcpp::Clock::SharedPtr clock, double rate)
+  : ROSRate(clock, rclcpp::Duration::from_seconds(1.0 / rate))
+  {
+  }
+  explicit ROSRate(rclcpp::Clock::SharedPtr clock, rclcpp::Duration period)
+  : clock_(clock), period_(period), last_interval_(clock->now())
+  {
+  }
+
+  virtual bool
+  sleep()
```

Comment:
I have a fundamental concern/worry is that we're conflating two concepts here. 

The rate object is a data storage which has an interval period and the last start time. 

The clock on the other hand is the thing that has a concept of time passing and keeps track of that. 

If you call XXXX.sleep() it has to be the thing that has a concept of time passing. Now we've gotten used to the ROS 1 Rate object being able to reach under the hood and grab the Node singleton and from that get the clock status and disambiguate sim time vs wall time and then give you an answer.  This implementation uses the original temporary rclcpp::sleep_for which was added before we had sim time and will not use that. 

However we've separated out these many layers so we now actually have a separate clock concept. And we could require that you construct a Rate object with the clock as proposed above. But that's tying together a much heaver concept which requires a time source and a lot more initialization. Whereas the Rate object is basically a little bit of data that doesn't need extra infrastructure. 

This can also be seen in Duration where we have not ported forward the `sleep` method that was previously defined in roscpp. And this parallels how chrono implements it. There's no Sleep defined on the "duration" object. https://en.cppreference.com/w/cpp/chrono/duration You call sleep_for on the duration because the Duration doesn't have a concept of time passing

Similarly for the Rate concept. You're going to call sleep_for(x) where x = start + period - now The Rate object doesn't need to know or handle time in the background etc. 
You can have and store Rate objects in messages and they stay valid without them additionally forcing association and maintenance of the clock object. Similarly different clocks could service a specific Rate object. 

To keep things isolated I would rather propose a Rate<T>.sleep(Clock<T>::ptr clock) Which will do the sleep and you pass it the appropriate clock when you want to sleep using it. This would be a shorthand for users to `if clock.sleep(Rate.get_remainder()) { Rate.reset() }` Maybe `sleep_remainder` or `sleep_cycle`

It would be valuable to enable the Rates to be templated on the Clock type to check that Rate<T> is compatible with Clock<T> so you don't end up sleeping from different timelines (Steady, Wall, ROS) 

Reply:
@tfoote, we also have the same need in Nav2 to use `rclcpp::Rate` respectful to ROS-time. If I undestood correctly, the main reason of the above - is to not overcomplicate the `rclcpp::Rate` by adding the references (and objects???) of `rclcpp::Clock` there?

We could template the `rclcpp::GenericRate` by `<rcl_clock_type_t Type>` instead of `<class Clock = std::chrono::clock>`. Depending on, templates' instantiations will be as follows:
```
using Rate = GenericRate<RCL_SYSTEM_TIME>;
using WallRate = GenericRate<RCL_STEADY_TIME>;
using ROSRate = GenericRate<RCL_ROS_TIME>;
```
Here I see 2 options of API solutions:

O1. Update `GenericRate::sleep(Clock::SharedPtr clock)` call to accept the pointer to RCLCPP Clock as input argument. Then inside `sleep()` implementation it will check whether given `clock` corresponds to a class template `<Type>` and use `clock->sleep_for(time_to_sleep)` API call to sleep.

O2. In the `GenericRate` constrictor, create an internal `rclcpp::Clock(Type)` object of the templated type and then store it in `GenericRate` privates: `clock_ = rclcpp::Clock::make_shared(Type);`. Then using `clock_->sleep_for()` inside the `sleep()`. This will free `sleep()` call from extra checks and developers from one more pointer to the clock in API. But it also will make a new `rclcpp::Clock` object inside `rclcpp::GenericRate`. Which if I understand is not the intention, if we want to `rclcpp::GenericRate` to be a lightweight structure. However, from my perspective, this option is pretty straight-forward to end-developers who will use this API.

Which way is better to prefer?
I may not see the whole picture, so please correct it if something missed or we could go another way.